src/scanner.c

```diff
@@ -0,0 +1,149 @@
+#include "tree_sitter/parser.h"
+#include <assert.h>
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+#define VEC_RESIZE(vec, _cap)                                                  \
+    void *tmp = realloc((vec).data, (_cap) * sizeof((vec).data[0]));           \
+    assert(tmp != NULL);                                                       \
+    (vec).data = tmp;                                                          \
+    (vec).cap = (_cap);
```

Comment:
Forgive me if this is obvious to someone that knows C, but why are these macros? Are you trying to avoid the function call overhead?

Reply:
It's really just preference, I took inspiration from Haskell's scanner and applied similar logic to upstream repos. I wouldn't mind if it were functions though.